acp/acp.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package acp
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/corelog"
+)
+
+var (
+	log = corelog.NewLogger("acp")
+
+	// NoACPModule is an empty ACPModule, this is used to disable access control.
+	NoACPModule = immutable.None[ACPModule]()
+)
+
+// ACPModule is the interface to all types of access control modules that might exist.
+type ACPModule interface {
+	// Init initializes the acp module, with an absolute path. The provided path indicates where the
+	// persistent data will be stored for the acp module.
+	Init(ctx context.Context, path immutable.Option[string])
```

Comment:
HAHA it's funny you bring that up, I actually didn't use `Option` before, and had it like you describe but was requested in review to change it to `Option`.

It'd be nice if @sourcenetwork/database-team we can have consensus here so we can stick with one choice.

Reply:
IMO this is a Go function, taking Go types, whilst `path` will typically be provided via a config file, it does not always have to be the case, and `Option` very strongly conveys that it is an optional value (both to callers of the function who may otherwise think it required, and to maintainers of the consuming code who may otherwise be unsure as to whether it will have a value or not).

Reply:
hmm, I'm mixed here, perhaps this is just my old school Go bias at play before Generics (and therefore before any reasonable implementation of Option) but I have fairly narrow use cases in my mind regarding Options, especially on interfaces.

As stated in my original comment, my issue stems from *how* the values are being interpreted. We *should* have two outcomes here 1) User has a defined path, and we use that path 2) User doesn't have a defined path and we use in-memory.

But in reality the Option gives us 3 options, that could be confusing. The two above, and 3) The user provides an empty value, and we error. Which may be confusing from a config value perspective, as that means to go back to the default,, you actually have to remove the `path` entry entirely from a user config, rather than just setting it to "empty". **If it does stay as is, then all 3 cases have to be clearly documented.**

Additionally, the Option is also forcing a specific kind of use-case on the interface, which is meant to be general and have various behaviors on the concrete implementation.. Some implementations could perfectly accept an empty string path for a variety of reasons.

> Option very strongly conveys that it is an optional value (both to callers of the function who may otherwise think it required, and to maintainers of the consuming code who may otherwise be unsure as to whether it will have a value or not).

This line suggests we are taking an opinionated stand for the implementors of this function, which doesnt feel right for the interface.

Reply:
To me both are fine, seems like a nitpick kind of thing, I slightly prefer string approach like I had it before, but happy to do the consensus.

@fredcarle @islamaliev preferences?

Reply:
I prefer the string approach here for the reasons John described.